content/300-guides/500-other/900-advanced-database-tasks/06-typesafe-raw-sql/01-safeql-gis.mdx

```diff
@@ -0,0 +1,128 @@
+---
+title: 'Using SafeQL with unsupported GIS features'
+metaTitle: 'Using SafeQL with unsupported GIS features (Guide)'
+metaDescription: 'Learn how to use SafeQL and Prisma Client extensions to work around features not natively supported by Prisma.'
+---
+
+## Overview
+
+This page explains how to use Prisma Client extensions and [SafeQL](https://safeql.dev) in order to make the experience of working with geographic data better.
+
+Our example will be using [PostGIS](https://postgis.net/) and PostgreSQL, but is applicable to other features and database engines.
+
+## Prerequisites
+
+To follow this guide, we recommend:
+
+- an existing [PostgreSQL](https://www.postgresql.org/) database
+- an existing Prisma project
+
+## What is SafeQL?
+
+[SafeQL](https://safeql.dev/) is an ESLint plugin that allows for advanced linting and type safety within raw SQL queries. After setup, SafeQL works with Prisma `$queryRaw` and `$executeRaw` to provide type safety when raw queries are required.
+
+## Geographic data support in Prisma
+
+At this time geographic data, and specifically PostGIS, are not supported by Prisma. Models that have geographic data columns will not have the benefit of type safety, nor will be able to use mutating operations like `create` and `update`.
+
+Prisma does support operations using `$queryRaw` and `$executeRaw` and we will be using these in order to create a Prisma Client extension with SafeQL to improve the usage of geographic data in Prisma.
+
+## 1. Set up Prisma for use with PostGIS
+
+If you haven't already, update your Prisma schema to add the `postgis` PostgreSQL extension
+
+```prisma
+generator client {
+  provider        = "prisma-client-js"
+  previewFeatures = ["postgresqlExtensions"]
+}
+
+datasource db {
+  provider   = "postgresql"
+  url        = env("DATABASE_URL")
+  extensions = [postgis]
+}
+```
+
+Then, create a migration using `npx prisma migrate dev` in order to enable the extension. The migration file should look like the following:
```

Comment:
```suggestion
Next, create a migration using the `prisma migrate dev` command to enable the extension. The output of the migration file should look like the following:
```

Reply:
Side note: I'm running into this error locally when using PostgreSQL 15 on Docker: 

```
❯ npx prisma migrate dev -n postgis-init
Prisma schema loaded from prisma/schema.prisma
Datasource "db": PostgreSQL database "safeql-gis-docs" at "localhost:5432"

PostgreSQL database safeql-gis-docs created at localhost:5432

Applying migration `20231106152917_postgis_init`
Error: P3018

A migration failed to apply. New migrations cannot be applied before the error is recovered from. Read more about how to resolve migration issues in a production database: https://pris.ly/d/migrate-resolve

Migration name: 20231106152917_postgis_init

Database error code: 0A000

Database error:
ERROR: extension "postgis" is not available
DETAIL: Could not open extension control file "/usr/share/postgresql/15/extension/postgis.control": No such file or directory.
HINT: The extension must first be installed on the system where PostgreSQL is running.

DbError { severity: "ERROR", parsed_severity: Some(Error), code: SqlState(E0A000), message: "extension \"postgis\" is not available", detail: Some("Could not open extension control file \"/usr/share/postgresql/15/extension/postgis.control\": No such file or directory."), hint: Some("The extension must first be installed on the system where PostgreSQL is running."), position: None, where_: None, schema: None, table: None, column: None, datatype: None, constraint: None, file: Some("extension.c"), line: Some(539), routine: Some("parse_extension_control_file") }
```

Is it expected to fail? I'm guessing someone trying this out the first time will likely run into this error too. 🤔

Reply:
I found this docker compose example in the prisma-examples we could use in a collapsible.
https://github.com/prisma/prisma-examples/blob/latest/typescript/postgis-express/docker-compose.yml


```
version: '3.6'
services:
  pgDB:
    image: postgis/postgis:13-3.1-alpine
    restart: always
    ports:
      - '5432:5432'
    volumes:
      - db_data:/var/lib/postgresql/data
    environment:
      POSTGRES_PASSWORD: password
      POSTGRES_DB: geoexample
volumes:
  db_data:
```